extendr-api/src/optional/faer.rs

```diff
@@ -0,0 +1,275 @@
+
+                if dim.len() != 2 {
+                    return Err("could not convert to matrix");
+                }
+
+                if let Some(slice) = robj.as_real_slice() {
+                    let fmat = mat::from_column_major_slice::<f64>(&slice, dim[0], dim[1]);
+                    Ok(fmat.to_owned())
+                } else if let Some(slice) = robj.as_integer_slice() {
+                    let fmat = Mat::<f64>::from_fn(dim[0], dim[1], |i, j| {
+                        if slice[i + j * dim[0]].is_na() {
+                            f64::NAN
+                        } else {
+                            slice[i + j * dim[0]] as f64
+                        }
+                    });
+                    Ok(fmat)
+                } else {
+                    Err("could not convert to matrix")
+                }
+            } else {
+                Err("could not convert to matrix")
+            }
+        } else {
+            Err("R object is not a matrix")
+        }
+    }
+}
+
+impl<'a> FromRobj<'a> for MatRef<'a, f64> {
+    fn from_robj(robj: &'a Robj) -> std::result::Result<Self, &'static str> {
+        if robj.is_matrix() {
+            if let Some(dim) = robj.dim() {
+                let dim: Vec<_> = dim.iter().map(|d| d.inner() as usize).collect();
+
+                if dim.len() != 2 {
+                    return Err("could not convert to matrix");
+                }
+
+                if let Some(slice) = robj.as_real_slice() {
+                    let fmat = mat::from_column_major_slice::<f64>(&slice, dim[0], dim[1]);
+                    Ok(fmat)
+                } else {
+                    Err("could not convert to matrix")
+                }
+            } else {
+                Err("could not convert to matrix")
+            }
+        } else {
+            Err("R object is not a matrix")
+        }
+    }
+}
+
+impl TryFrom<&Robj> for Mat<f64> {
+    type Error = Error;
+
+    fn try_from(robj: &Robj) -> Result<Self> {
+        if robj.is_matrix() {
+            if let Some(dim) = robj.dim() {
+                let dim: Vec<_> = dim.iter().map(|d| d.inner() as usize).collect();
+
+                if dim.len() != 2 {
+                    return Err(Error::ExpectedMatrix(robj.clone()));
+                }
+
+                if let Some(slice) = robj.as_real_slice() {
+                    let fmat = mat::from_column_major_slice::<f64>(&slice, dim[0], dim[1]);
+                    Ok(fmat.to_owned())
+                } else {
+                    Err(Error::ExpectedReal(robj.clone()))
+                }
+            } else {
+                Err(Error::ExpectedMatrix(robj.clone()))
+            }
+        } else {
+            Err(Error::ExpectedMatrix(robj.clone()))
+        }
+    }
+}
+
+impl<'a> TryFrom<&'_ Robj> for MatRef<'a, f64> {
+    type Error = Error;
+
+    fn try_from(robj: &Robj) -> Result<Self> {
+        if robj.is_matrix() {
```

Comment:
I've recently checked the behavior of `isMatrix` and it checks a lot of these things at once:
```c
INLINE_FUN Rboolean isMatrix(SEXP s)
{
    SEXP t;
    if (isVector(s)) {
	t = getAttrib(s, R_DimSymbol);
	/* You are not supposed to be able to assign a non-integer dim,
	   although this might be possible by misuse of ATTRIB. */
	if (TYPEOF(t) == INTSXP && LENGTH(t) == 2)
	    return TRUE;
    }
    return FALSE;
}
```

Reply:
Nice, I think https://github.com/extendr/extendr/pull/706/commits/4995ebbdc699182840767eb3e98137f0d86f8f62 addresses the unnecessary check. That said would it be possible to change this line `let dim: Vec<_> = dim.iter().map(|d| d.inner() as usize).collect();`  and use the dimension directly? Maybe with [as_array_view](https://docs.rs/extendr-api/0.6.0/extendr_api/prelude/trait.Dimension.html#method.as_array_view) ?

Reply:
I recently went through this, you can see it here https://github.com/CGMossa/extendr/blob/master/extendr-api/src/optional/ndarray.rs

Reply:
Sorry that was the wrong reference..
```rs

impl<T> TryFrom<&Robj> for RMatrix<T>
where
    Robj: for<'a> AsTypedSlice<'a, T>,
{
    type Error = Error;

    fn try_from(robj: &Robj) -> Result<Self> {
        // isMatrix checks vector-type, presence of `dim`, and length(dim) == 2
        if !robj.is_matrix() {
            return Err(Error::ExpectedMatrix(robj.clone()));
        }
        let dim = robj.dim().unwrap();
        let dim = dim.as_slice();
        Ok(RArray::from_parts(
            robj.clone(),
            [dim[0].inner() as _, dim[1].inner() as _],
        ))
    }
}
```

Reply:
Turns out, `.as_slice()` isn't in `master`.